### PR TITLE
Meta no longer clones - fixes #233

### LINKED ISSUE #233: Resource unnecessarily clones
The [`Meta` helper trait](https://docs.rs/kube/0.32.1/kube/api/trait.Meta.html) has the following signatures,

```
fn name(&self) -> String
fn namespace(&self) -> Option<String>
fn resource_ver(&self) -> Option<String>
```

All of these end up cloning the underlying thing they're returning; none of them need to; they could return `&str` references.

In my particular case, I'm iterating through every object of a particular resource type, this results in quite a few allocations that I end up never needing. (I'm just comparing the `name`, for example.)

This would be a breaking change, however, since `kube-rs` isn't at 1.0 yet, I thought I'd suggest it.

### PR DESCRIPTION
**Pull request by clux**
As this is now causing more congestion on shared references (as seen in the examples), it makes it awkward when used with option heavy k8s_openapi.

If k8s_openapi implements https://github.com/Arnavion/k8s-openapi/issues/72 then this will feel a lot less awkward and should almost certainly be merged then.

At any rate, wanted to see how this felt. Going to leave this one open for discussion for a bit.

### CODE REVIEW

File: examples/crd_api.rs [hunk 1]
```diff
@@ -50,7 +50,7 @@ async fn main() -> anyhow::Result<()> {
             info!(
                 "Deleting {}: ({:?})",
                 Meta::name(&o),
-                o.status.unwrap().conditions.unwrap().last()
+                o.status.as_ref().unwrap().conditions.as_ref().unwrap().last()
```
Comment (clux) [inline examples/crd_api.rs:53]: This is the kind of awkwardness I am talking about ^

File: kube/src/api/metadata.rs [hunk 2]
```diff
@@ -11,17 +11,15 @@ use serde::{Deserialize, Serialize};
...
     fn meta(&self) -> &ObjectMeta;
     /// The name of the resource
-    fn name(&self) -> String;
+    fn name(&self) -> &String;
     /// The namespace the resource is in
-    fn namespace(&self) -> Option<String>;
+    fn namespace(&self) -> &Option<String>;
     /// Tthe resource version
-    fn resource_ver(&self) -> Option<String>;
+    fn resource_ver(&self) -> &Option<String>;
```
Comment (teozkr) [inline kube/src/api/metadata.rs:22]: Should be `&str` and `Option<&str>`, respectively. That should cut down on the awkward `as_ref()`s, and `Option<&T>` is covered by the null pointer optimization.

Comment (clux) [reply kube/src/api/metadata.rs]: Pushed a change now that tries to do that signature. It shifts things a lot, but it doesn't feel any less awkward. If anything it feels more awkward internally.

File: kube/src/api/metadata.rs [hunk 3]
```diff
@@ -33,16 +31,16 @@ where
...
-    fn resource_ver(&self) -> Option<String> {
-        self.meta().resource_version.clone()
+    fn resource_ver(&self) -> Option<&str> {
+        self.meta().resource_version.as_ref().map(|rv| rv.as_ref())
     }
 
-    fn namespace(&self) -> Option<String> {
-        self.meta().namespace.clone()
+    fn namespace(&self) -> Option<&str> {
+        self.meta().namespace.as_ref().map(|ns| ns.as_ref())
```
Comment (clux) [inline kube/src/api/metadata.rs:43]: `&Option<String>` vs `Option<&str>` awkwardness ^

Comment (teozkr) [reply kube/src/api/metadata.rs:43]: Should be `self.meta().namespace.as_deref()`

File: examples/configmap_reflector.rs [hunk 4]
```diff
@@ -12,7 +12,11 @@ fn spawn_periodic_reader(reader: Store<ConfigMap>) {
         loop {
             // Periodically read our state
             tokio::time::delay_for(std::time::Duration::from_secs(5)).await;
-            let cms: Vec<_> = reader.state().iter().map(|obj| Meta::name(obj).clone()).collect();
+            let cms: Vec<_> = reader
+                .state()
+                .iter()
+                .map(|obj| Meta::name(obj).to_string())
```
Comment (teozkr) [inline examples/configmap_reflector.rs:18]: Think this should be able to borrow it? So just `.map(Meta::name)`.

File: kube/src/api/metadata.rs (hunk 3 above)
Comment (clux) [reply kube/src/api/metadata.rs]: That is nicer. Thanks!

File: examples/configmap_reflector.rs (hunk 4 above)
Comment (clux) [reply examples/configmap_reflector.rs:18]: No, that doesn't work unfortunately. "Temporary value dropped while burrowed".

Comment (teozkr) [reply examples/configmap_reflector.rs:18]: Oh, aw. I guess something is off with our lifetime bounds somewhere? :/

Comment (clux) [reply examples/configmap_reflector.rs:18]: I think this is just the standard issue where we are trying to collect references to strings that are only available in the scope up. `reader` is dropped at the end of `collect`, and then you just have a vector of dangling slices.

Comment (teozkr) [reply examples/configmap_reflector.rs:18]: `reader` is still valid, but now that you mention it, it is correct that the handle expires with the `Iterator`.. :/

Comment (clux) [reply examples/configmap_reflector.rs:18]: Ah, yeah, sorry. the `state()` temporary value is dropped.

Comment (clux) [reply examples/configmap_reflector.rs]: Still nice enough to use by caching `reader.state()` done now.

**Pull request closed without merging**